### ISSUE TITLE
C# exported game doesn't work on Linux, missing Mono data

### ISSUE THREAD

**Issue by Jcozman** [godotengine/godot#25449]
**Godot version:**
3.1 beta 3

**OS/device including version:**
Ubuntu 18.04.1

**Issue description:**
In preparation for a future Steam release, Steam staff encountered the following crash:

```
On Ubuntu 18.04.1 LTS, the app crashes as soon as a class is selected. On SteamOS, the app crashes after the starting window appears. 

The same crash occurs after clicking the executable file in the install directory.

The following error is shown in the console:

"ERROR: ld.so: object '/home/gamer/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
>>> Adding process 6389 for game ID 906450
[New LWP 6362]
[New LWP 6363]
[New LWP 6364]
[New LWP 6365]
[New LWP 6366]
[New LWP 6367]
>>> Adding process 6390 for game ID 906450
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f68e6ea723a in __waitpid (pid=pid@entry=6389, stat_loc=stat_loc@entry=0x7fff1d7dc82c, options=options@entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
30	../sysdeps/unix/sysv/linux/waitpid.c: No such file or directory.
Id Target Id Frame 
* 1 Thread 0x7f68e74ff880 (LWP 6361) "Tenacious.x86_6" 0x00007f68e6ea723a in __waitpid (pid=pid@entry=6389, stat_loc=stat_loc@entry=0x7fff1d7dc82c, options=options@entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
2 Thread 0x7f68dd381700 (LWP 6362) "Tenacious.x86_6" 0x00007f68e6ea56d6 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x4c0c348) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
3 Thread 0x7f68dcc69700 (LWP 6363) "Tenacious.x86_6" 0x00007f68e6ea56d6 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x4fb2c48) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
4 Thread 0x7f68c6c8f700 (LWP 6364) "Tenacious.x86_6" 0x00000000020f42d0 in String::operator==(String const&) const ()
5 Thread 0x7f68c6c4e700 (LWP 6365) "Tenacious.x86_6" 0x00007f68e63de9d0 in __GI___nanosleep (requested_time=requested_time@entry=0x7f68c6c4dcd0, remaining=remaining@entry=0x0) at ../sysdeps/unix/sysv/linux/nanosleep.c:28
6 Thread 0x7f68c63ff700 (LWP 6366) "SGen worker" 0x00007f68e6ea29f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x2bac828 <work_cond+40>) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
7 Thread 0x7f68c4bfe700 (LWP 6367) "Finalizer" 0x00007f68e6ea56d6 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x2b9a880 <finalizer_sem>) at ../sysdeps/unix/sysv/linux/futex-internal.h:205

[...]

=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application."
```

Additionally, in the `mono_logs` folder there is a single file with a single line:
```
cant resolve internal call to "Interop/Sys::DoubleToString" (tested without signature also) (in domain , warning)
```

**Steps to reproduce:**
I'm unsure of the steps to reproduce, I lack a Linux environment to test on at the moment. It looks like a Mono issue. I can share the exported game files on request, but I don't want to share the full game here (for obvious reasons).

Let me know if this is a valid issue. If not I will close, but would appreciate some insight into what's going on here.

**Minimal reproduction project:**

**Comment by akien-mga** [comment]
Can you still reproduce the issue with 3.1 beta 8?

If so, can you give details on how you packaged your Linux build for Steam? In particular, did you make sure to include the data folder containing C# assemblies that should be exported together with your Linux binary and .pck?

**Comment by Jcozman** [comment]
I don't recall a `data` folder being created with the exported binary and pck. I built for Linux using the Windows build of Godot with mono support (beta 3) and I uploaded only the binary and pck to Steam because those were the only generated files. I'm away from my development computer for the next few days but I will test it out on the latest beta and get back to you when I can.

**Comment by akien-mga** [comment]
I tested with 3.1 beta 6 Mono, and I confirm that no data folder was exported, and I got this error in the terminal:
```
Unhandled Exception:
System.IO.FileNotFoundException: Data template directory not found
  at GodotSharpTools.Editor.GodotSharpExport._ExportBegin (System.String[] features, System.Boolean debug, System.String path, System.Int32 flags) [0x000c3] in <c3891e5a8dad40e2a99fcc2bedb221ab>:0 
ERROR: _export_begin: Method/Function Failed.
   At: modules/mono/editor/godotsharp_export.cpp:163.
```

I'll try 3.1 beta 8 now. CC @hpvb

**Comment by akien-mga** [comment]
I tested with 3.1 beta 8, and this time the data folder is properly exported.

Yet, the game still doesn't work, even on the same system that was used to export the binary (Mageia 7 x86_64):
```
$ ls -l
total 67304
drwxr-xr-x 3 akien akien     4096 Feb 28 10:45 data_InteractiveCubes/
-rw-r--r-- 1 akien akien   516980 Feb 28 10:45 multicube_mono_debug.pck
-rwxr-xr-x 1 akien akien 68386415 Feb 28 10:45 multicube_mono_debug.x86_64*

$ ./multicube_mono_debug.x86_64 
OpenGL ES 2.0 Renderer: GeForce GTX 670MX/PCIe/SSE2
The assembly mscorlib.dll was not found or could not be loaded.
It should have been installed in the `/home/akien/tmp/godot/bug/multicube_mono_linux/debug/data_InteractiveCubes/Mono/lib/mono/4.5/mscorlib.dll' directory.

$ find data_InteractiveCubes/  
data_InteractiveCubes/
data_InteractiveCubes/Mono
data_InteractiveCubes/Mono/etc
data_InteractiveCubes/Mono/etc/mono
data_InteractiveCubes/Mono/etc/mono/4.5
data_InteractiveCubes/Mono/etc/mono/4.5/DefaultWsdlHelpGenerator.aspx
data_InteractiveCubes/Mono/etc/mono/4.5/settings.map
data_InteractiveCubes/Mono/etc/mono/4.5/machine.config
data_InteractiveCubes/Mono/etc/mono/4.5/Browsers
data_InteractiveCubes/Mono/etc/mono/4.5/Browsers/Compat.browser
data_InteractiveCubes/Mono/etc/mono/4.5/web.config
data_InteractiveCubes/Mono/etc/mono/mconfig
data_InteractiveCubes/Mono/etc/mono/mconfig/config.xml
data_InteractiveCubes/Mono/etc/mono/4.0
data_InteractiveCubes/Mono/etc/mono/4.0/DefaultWsdlHelpGenerator.aspx
data_InteractiveCubes/Mono/etc/mono/4.0/settings.map
data_InteractiveCubes/Mono/etc/mono/4.0/machine.config
data_InteractiveCubes/Mono/etc/mono/4.0/Browsers
data_InteractiveCubes/Mono/etc/mono/4.0/Browsers/Compat.browser
data_InteractiveCubes/Mono/etc/mono/4.0/web.config
data_InteractiveCubes/Mono/etc/mono/2.0
data_InteractiveCubes/Mono/etc/mono/2.0/DefaultWsdlHelpGenerator.aspx
data_InteractiveCubes/Mono/etc/mono/2.0/settings.map
data_InteractiveCubes/Mono/etc/mono/2.0/machine.config
data_InteractiveCubes/Mono/etc/mono/2.0/Browsers
data_InteractiveCubes/Mono/etc/mono/2.0/Browsers/Compat.browser
data_InteractiveCubes/Mono/etc/mono/2.0/web.config
data_InteractiveCubes/Mono/etc/mono/browscap.ini
data_InteractiveCubes/Mono/etc/mono/config
data_InteractiveCubes/Mono/lib
data_InteractiveCubes/Mono/lib/libMonoPosixHelper.so
data_InteractiveCubes/Mono/lib/libmono-btls-shared.so
```

Some stuff is missing in the data folder.

**Comment by akien-mga** [comment]
The above test was made with a test project with no C# solution at all (and no C# code).

Now I've generated a C# solution in the same project, and when exporting a debug build, I get this error:
```
ERROR: _add_file: Condition ' !f ' is true. returned: false
   At: modules/mono/editor/godotsharp_export.cpp:170.
ERROR: _export_begin: Condition ' !_add_file(scripts_metadata_path, scripts_metadata_path) ' is true.
   At: modules/mono/editor/godotsharp_export.cpp:95.
```
And this time, no data folder at all. (Still 3.1 beta 8.)

That seems to happen because I generated the solution but never built it.


Next step, I added a C# script and build the solution. This time the export worked fine, and I get a functional game in the end. The data folder still has the same contents, but Mono libs are now included in the .pck file:
```
  inflating: .mono/metadata/scripts_metadata.debug  
  inflating: .mono/assemblies/GodotSharp.dll  
  inflating: .mono/assemblies/InteractiveCubes.dll  
  inflating: .mono/assemblies/Mono.Security.dll  
  inflating: .mono/assemblies/System.dll  
  inflating: .mono/assemblies/System.Configuration.dll  
  inflating: .mono/assemblies/System.Core.dll  
  inflating: .mono/assemblies/System.Numerics.dll  
  inflating: .mono/assemblies/System.Security.dll  
  inflating: .mono/assemblies/System.Xml.dll  
  inflating: .mono/assemblies/mscorlib.dll
```

**Comment by akien-mga** [comment]
TL;DR: False alarm, as of 3.1 beta 8 expecting *valid* C# projects works on Linux, and should work for @Jcozman's game.

The two other bugs I found happened when you don't have a C# solution, or when you didn't build it, but they're minor issues and I'll open separate bug reports for them.

**Comment by akien-mga** [comment]
@Jcozman Would still be great if you can confirm that your game works when exported with 3.1 beta 8 (or beta 9 which should be released today). I can test a Linux build for you if needed.

**Comment by Jcozman** [comment]
@akien-mga Just tested on beta 10, the Linux build works perfectly fine now. Thank you!

**Comment by akien-mga** [comment]
Nice! I see you added a Linux build for the game on Steam, I'll buy it :D
You still need to enable Linux in Steamworks though, it doesn't appear with the Linux / SteamOS logo in the store.

**Comment by Jcozman** [comment]
Thank you! I'll take a look - it could be that that the store page hasn't updated yet. I was able to download and run the game from the Steam client in my virtual machine running Ubuntu.
